### ISSUE TITLE
Bug: lsp_definition jumps are not possible with denols.

### ISSUE THREAD

**Issue by uga-rosa** [ibhagwan/fzf-lua#1305]
### RTFM Checklist

- [X] I have searched exisiting issues / discussions
- [X] I have read the Wiki including the Advanced section
- [X] I have read `man fzf` / I am well versed in shell fzf

### Operating system

Ubuntu (WSL2)

### Shell

fish

### Neovim version (`nvim --version`)

nvim v0.10.0

### Fzf version (`fzf --version`)

0.52.1 (6432f00)

### Output of `:lua print(os.getenv('FZF_DEFAULT_OPTS'))`

--height 50% --reverse

### Is the problem reproducible with `mini.sh`?

- [ ] My issue is reproducible with `mini.sh`
- [ ] My issue **IS NOT** reproducible with `mini.sh`
- [x] I have not tested with `mini.sh` (not relevant, requires LSP, Windows, etc)

### Fzf-lua configuration

```lua
require('fzf-lua').setup({
})
```

### Describe the bug / steps to reproduce

I know the cause.

`deno ls` must create a virtual buffer in advance to jump to the library definition.
This process is registered in client.handlers in nvim-lspconfig. `vim.lsp.buf.definition` calls it so the definition jump can be performed correctly.

- set custom handler

https://github.com/neovim/nvim-lspconfig/blob/1d2454f9fbb34206897b9a1b5e8b87d6418941ba/lua/lspconfig/server_configurations/denols.lua#L91

- call custom handler

https://github.com/neovim/neovim/blob/46187117c9464dd6c09eae8e47bb279d72a01038/runtime/lua/vim/lsp/buf.lua#L44

In addition to deno ls, there may be other cases that require special ls-specific processing, and mechanisms already exist to cover them. So, why not make fzf-lua use this as well?

**Comment by ibhagwan** [comment]
Not sure I understand, fzf-lua calls `vim.lsp.util.jump_to_location` directly and let’s neovim handle the rest:
https://github.com/ibhagwan/fzf-lua/blob/41ea49cfc1b3f0586753bbeb7850474f28dc22cf/lua/fzf-lua/actions.lua#L172

When and why should we be calling the handler? What is not working with deno ls?

**Comment by uga-rosa** [comment]
My English is not very good, so sorry if I didn't convey it well...

`vim.lsp.buf.definition()` send request and pass response to handler.
deno ls replaces this handler to achieve a custom hook.
In this handler, the `util.jump_to_location()` is called.
In other words, simply calling `util.jump_to_location()` directly does not call the replaced custom hook.

- default definition handler

https://github.com/neovim/neovim/blob/46187117c9464dd6c09eae8e47bb279d72a01038/runtime/lua/vim/lsp/handlers.lua#L459

**Comment by ibhagwan** [comment]
I'm really not sure what I'm supposed to do with this information, fzf-lua isn't supposed to replace the handler, if I want definitions I call `:FzfLua lsp_definitions`, I don't replace the handler and call `vim.lsp.buf.definition()` as there's no point, if I really wanted to replace the same workflow I might map `gd -> :FzfLua lsp_definitions`, it's not fzf-lua's job to take into account every LSP's implementaiton and custom handlers.

**Comment by ibhagwan** [comment]
if you mean that `vim.lsp.util.jump_to_location()` is failing because some handler isn't called I also believe that should be handled upstream, the point of `util.jump_to_location()` is exactly that, to jump to an LSP location, if that fails because of custom handlers unknown to the caller (user) that should be handled in neovim or nvim-lspconfig, not in fzf-lua.

**Comment by uga-rosa** [comment]
Ah, I see, I understand. This is indeed difficult to solve in fzf-lua itself.
I will try to solve it here.
Thanks for your reply!

**Comment by ibhagwan** [comment]
From `:help jump_to_location`:
```vimdoc
                                             *vim.lsp.util.jump_to_location()*
jump_to_location({location}, {offset_encoding}, {reuse_win})
    Jumps to a location.

    Parameters: ~
      • {location}         (`table`) (`Location`|`LocationLink`)
      • {offset_encoding}  (`string?`) utf-8|utf-16|utf-32
      • {reuse_win}        (`boolean?`) Jump to existing window if buffer is
                           already open.

    Return: ~
        (`boolean`) `true` if the jump succeeded
```

Nothing in the help mentions requiring additional calls to custom handlers, if the first call fails due to a specialized handler not called, based on the links you posted this would be a neovim lua runtime issue which should be handled upstream.

**Comment by uga-rosa** [comment]
I leave this as I have investigated the issue in more detail and found a solution.

It was not vim.lsp.util.jump_to_location that was the problem.
Before executing `vim.lsp.util.locations_to_items`, the virtual buffer to jump to must be created or an error will occur.

Therefore, this can be solved in this way.

```lua
local function virtual_text_document(uri)
  local bufnr = vim.uri_to_bufnr(uri)
  local current_buf = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
  if #current_buf ~= 0 then
    return nil
  end
                                                                      
  local method = "deno/virtualTextDocument"
  local params = { textDocument = { uri = uri } }
  local client = vim.lsp.get_clients({ name = "denols" })[1]
  local result = client.request_sync(method, params, 2000, 0)
  if not result or type(result.result) ~= "string" then
    return nil
  end
  local lines = vim.split(result.result, "\n")
                                                                      
  vim.api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
  vim.api.nvim_set_option_value("readonly", true, { buf = bufnr })
  vim.api.nvim_set_option_value("modified", false, { buf = bufnr })
  vim.api.nvim_set_option_value("modifiable", false, { buf = bufnr })
  vim.lsp.buf_attach_client(bufnr, client.id)
end
                                                                      
vim.lsp.util.locations_to_items = (function(wrapped)
  return function(locations, encoding)
    for _, loc in ipairs(locations) do
      local uri = loc.uri or loc.targetUri
      if vim.startswith(uri, "deno:/") then
        virtual_text_document(uri)
      end
    end
    return wrapped(locations, encoding)
  end
end)(vim.lsp.util.locations_to_items)
```

**Comment by ibhagwan** [comment]
Shouldn’t `vim.lsp.util.locations_to_items` work without your solution? Maybe it’s worth to upstream this? Uses calling `vim.lsp.util.locations_to_items` shouldn’t be aware of LSP server details like this.

**Comment by uga-rosa** [comment]
I'll try to open an issue about this upstream tomorrow, but I honestly don't expect it to work.
I don't think they designed vim.lsp.util as something users touch directly, so I think they will just tell me to use vim.lsp.buf.definition and that's it...

**Comment by ibhagwan** [comment]
> I'll try to open an issue about this upstream tomorrow, but I honestly don't expect it to work. I don't think they designed vim.lsp.util as something users touch directly, so I think they will just tell me to use vim.lsp.buf.definition and that's it...

I doubt it, it’s an API with `:help` and as such it should work without the knowledge of specific LSP details.

**Comment by uga-rosa** [comment]
The help does not describe what lsp.Location is, etc., and I don't think it is a polite help that can be used without knowledge of lsp. I think it's a simple explanation for plugin developers who have knowledge, not for end users.

However, it would be worthwhile to raise this report upstream, as I have confirmed that the same problem occurs with telescope.nvim. The current system is not very good.

BTW, I noticed that vim.lsp.buf.definition has an option called on_list. This is a callback for multiple results, and if I could get fzf-lua to fire up here, it might work with the custom handler in lspconfig.
However, when jump_to_single_result is not enabled, fzf-lua must be opened even for single candidates, so this is just a thought and not a practical solution.
